contracts/02-tamagotchi-interaction/src/lib.rs

```diff
@@ -24,17 +70,73 @@ extern fn handle() {
-            let age = exec::block_timestamp() - tmg.date_of_birth;
-            msg::reply(TmgEvent::Age(age), 0).expect("Error in a reply'tamagotchi::age'");
+    if msg::source() == tmg.owner {
+        match action {
+            TmgAction::Name => {
+                msg::reply(TmgEvent::Name(tmg.name.clone()), 0)
+                    .expect("Error in a reply'tamagotchi::name'");
+            }
+            TmgAction::Age => {
+                let age = exec::block_timestamp() - tmg.date_of_birth;
+                msg::reply(TmgEvent::Age(age), 0).expect("Error in a reply'tamagotchi::age'");
+            }
+            TmgAction::Feed => {
+                if tmg.current_fed() <= 9000 {
+                    let fed = tmg.fed + FILL_PER_FEED;
+                    msg::reply(TmgEvent::Fed, 0).expect("Error in a reply'tamagotchi::fed'");
+                    tmg.fed = fed;
+                    tmg.fed_block = exec::block_height() as u64;
+                    tmg.entertained = tmg.current_entertained();
+                    tmg.slept = tmg.current_slept();
+                } else {
+                    let fedblock = exec::block_height() as u64;
+                    tmg.fed = 10000;
+                    tmg.fed_block = fedblock;
+                    tmg.entertained = tmg.current_entertained();
+                    tmg.slept = tmg.current_slept();
+                    msg::reply(TmgEvent::Fed, 1).expect("Error in a reply'tamagotchi::fed'");
+                }
+            }
+            TmgAction::Entertain => {
+                if tmg.current_entertained() <= 9000 {
+                    let entertained = tmg.entertained + FILL_PER_ENTERTAINMENT;
+                    msg::reply(TmgEvent::Entertained, 0)
+                        .expect("Error in a reply'tamagotchi::entertained'");
+                    tmg.entertained = entertained;
+                    tmg.entertained_block = exec::block_height() as u64;
+                    tmg.fed = tmg.current_fed();
+                    tmg.slept = tmg.current_slept();
+                } else {
+                    let entertainedblock = exec::block_height() as u64;
+                    tmg.entertained = 10000;
+                    tmg.entertained_block = entertainedblock;
+                    tmg.fed = tmg.current_fed();
+                    tmg.slept = tmg.current_slept();
+                    msg::reply(TmgEvent::Entertained, 1)
+                        .expect("Error in a reply'tamagotchi::entertained'");
+                }
+            }
+            TmgAction::Sleep => {
+                if tmg.current_slept() <= 9000 {
+                    let slept = tmg.slept + FILL_PER_SLEEP;
+                    msg::reply(TmgEvent::Slept, 0).expect("Error in a reply'tamagotchi::slept'");
+                    tmg.slept = slept;
+                    tmg.slept_block = exec::block_height() as u64;
+                    tmg.fed = tmg.current_fed();
+                    tmg.entertained = tmg.current_entertained();
```

Comment:
Anyway, you can move the `fed` / `entertained` / `slept` decreasing outside this `match`.